clang-tools-extra/clang-tidy/bsl/TypeDeclHeaderCheck.cpp

```diff
@@ -0,0 +1,71 @@
+//===--- TypeDeclHeaderCheck.cpp - clang-tidy -----------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TypeDeclHeaderCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace bsl {
+
+void TypeDeclHeaderCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(tagDecl().bind("x"), this);
```

Comment:
Should this be namedDecl instead of tagDecl?

Reply:
I'm trying to match classes, structs, enums, and unions with this. I'll need to add typedefs though